### ISSUE TITLE
support running without qdarkstyle for better Linux Desktop integration

### ISSUE THREAD

**Issue by sten0** [borgbase/vorta#471]
This request is related to #331

After working on packaging Vorta and QDarkStyle, I realised that defaulting to a non-system Qt style is not what most users want.  Namely, for a package installed with the system package manager on Debian (and derivatives such as Ubuntu, Kubuntu, Neon, etc) the expectation is that all applications follow the style of the desktop as a whole.  This is from a Qt/KDE Plasma perspective; 

Pyqt5 provides access to the default style, whatever it may be.  I'm not sure what the ideal solution is, because QDarkStyle is so different from the usual way of doing things on the Linux desktop, but I'm guessing it will be something like boolean somewhere in Vorta's source.  When toggled (with a patch or with sed, when the deb is created), Vorta would not attempt to import QDarkStyle, would disable in-application theme switching (GUI), and would skip all code paths involving QDarkStyle, instead using the pyqt5 defaults.

**Comment by Hofer-Julian** [comment]
I personally would be fine with removing it all together. Not sure how it looks on macOS, but on Linux it looks terrible and we got quite a few bug reports because of it.

**Comment by m3nu** [comment]
I'm fine with the defaults too.

Will remove it today and see if bad things happen.

**Comment by m3nu** [comment]
I'm done removing all custom styles, @Hofer-Julian @sten0 . Works well on macOS. Even changes in real-time now. Was able to get rid of a few workarounds for icon colors as well.

Please check it out for Linux. Especially the function [`uses_dark_mode()`](https://github.com/borgbase/vorta/blob/a5aaf12a2797da001a8f7842c41c02bdad67e920/src/vorta/utils.py#L212) should be adapted for Linux. Then we can also get rid of some settings. Hope Qt nicely adjusts to to the main color there.

**Comment by Hofer-Julian** [comment]
Will look into it this weekend

**Comment by sten0** [comment]
Here is a screencap from my WIP Debian package, with #475 merged, running in a sid VM:

[screencap of PR for 471.zip (1.5M)](https://github.com/borgbase/vorta/files/4696738/screencap.of.PR.for.471.zip)

**Comment by m3nu** [comment]
Thanks for the screencap, @sten0 . Looks surprisingly similar to the Fusion theme. Sure it's merged? But then it adjusted when you changed the theme, which is good.

I do see we have some remaining design issues. Generally we only use plain-vanilla Qt, so most elements should adjust well.

I did have to set up an event listener, looking for theme changes. This seems to work on Linux as well, but the `uses_dark_theme()` isn't implemented for Linux. Not sure where to get this info.

The issues you have with the `QToolbox` element (the horizontal tabs) also exist on macOS, but less severe. Just have the wrong background. Closing a re-opening the window fixed it, but we could probably repaint the element manually when the theme change fires.

To summarize the tasks I see ahead:

- [ ] Detect light/dark theme on Linux `uses_dark_theme()` (this will also fix the icons)
- [ ] Fix `QToolbox` display after theme change without closing window.
- [ ] Fix `QToolbox` titles (I think this is why we used the Fusion style in the first place)

I'll probably set up a quick VM as well. Is this just the default desktop?

**Comment by Hofer-Julian** [comment]
> I did have to set up an event listener, looking for theme changes. This seems to work on Linux as well, but the `uses_dark_theme()` isn't implemented for Linux. Not sure where to get this info.

That's a thing of the future, unfortunately.
https://blog.elementary.io/the-need-for-a-freedesktop-dark-style-preference/

**Comment by m3nu** [comment]
Quick note: It's possible to get the UI colors via the palette attribute. This should enable system-independent dark mode detection. E.g. to get the text color printed, plug this in near `vorta/application.py:47`

```
        palette = self.tray.contextMenu().palette()
        print(palette.window().color().name())
        print(palette.windowText().color().name())
```

For dark theme:
```
#323232
#ffffff
```

Light theme:
```
#ececec
#000000
```

**Comment by m3nu** [comment]
> That's a thing of the future, unfortunately.

No problem. We can directly get the colors from Qt. Just needs some threshold of what's dark.

**Comment by sten0** [comment]
> Thanks for the screencap, @sten0 . Looks surprisingly similar to the Fusion theme. Sure it's merged? But then it adjusted when you changed the theme, which is good.
>

You're welcome @m3nu, and thank you for working on this issue :-) This is
the default theme for KDE Plasma.  The light variant is called
"Breeze" and the dark variant "Breeze Dark".

> I did have to set up an event listener, looking for theme changes. This seems to work on Linux as well, but the `uses_dark_theme()` isn't implemented for Linux. Not sure where to get this info.
>

More on this later.

> To summarize the tasks I see ahead:
> 
> * [ ]  Detect light/dark theme on Linux `uses_dark_theme()` (this will also fix the icons)
> * [ ]  Fix `QToolbox` display after theme change without closing window.
> * [ ]  Fix `QToolbox` titles (I think this is why we used the Fusion style in the first place)
>

Yes, I think that's everything I hoped to say with the screencap :-)

> I'll probably set up a quick VM as well. Is this just the default desktop?

Yes.  Steps to get this specific mix of packages:

1. Start with a stable minimal ISO or small cd.  Install nothing except the standard tools.
2. Install it.  To disable the root account and use sudo like on Ubuntu, take care to leave the root password blank--I'm not sure if this was ever documented!
3. Reboot, log in, and replace /etc/apt/sources.list with the following:
```
deb http://deb.debian.org/debian unstable main contrib non-free
#  /\ or your preferred mirror /\
```
4. `sudo apt update` and `sudo apt full-upgrade`
5. Reboot. (use the `sudo reboot` command)
6. Log in and run `sudo apt full-upgrade` again for good measure.
6a. The unstable distribution/suite is changing constantly, but not everything moves at the same speed.  Force installations and removals will expose consistency problems that require advanced skills to recover from.
7. `sudo apt install kde-plasma-desktop`
7.a: Either reboot or run `sudo systemctl start sddm`
8. What about system-integrated deps?

I'm assuming you'll be working from git and skipping the work of Debian packaging.  At a minimum `python3-pyqt5.qsci python3-secretstorage` should be installed.  The following is the list of build-dependencies I use:
```
python3-all
python3-setuptools
python3-setuptools-git
python3-pytest-forked
python3-pytest-runner
python3-appdirs
python3-apscheduler
python3-keyring
python3-dateutil
python3-paramiko
python3-peewee
python3-psutil
python3-pyqt5.qsci
```

IIRC I had to manually specify `python3-pyqt5.qsci python3-secretstorage` as runtime dependencies, and the rest were picked up automatically from the build using the dh-python (and should probably be installed).

And that's my new VM testbed.

I wasn't yet able to get Vorta to work on Debian 10 (buster); here there was failed to import pyqt5 error (low priority for me, since I'm targeting Debian 11).  A Neon LTS based on Ubuntu 20.04 might be new enough though.  No idea about Kubuntu.  They'll all inherit the eventual Debian package their future versions, so the number of your users will grow quite a bit :-)  'hope that's what you want!

**Comment by m3nu** [comment]
Thanks for the VM instructions. I was afraid to ask 😄. Will try that this morning.

The other items are probably taken care of already. Detecting dark mode based on palette colors works perfectly and the style issue in `QToolbox` was due to a bold style I applied. After I removed it, it worked flawlessly. This may fix the titles as well, since the style was applied there too. (just to make it bold, but seems to confuse Qt)

I'll check back after testing on the VM.

**Comment by m3nu** [comment]
Looks like all the issues are already solved. Not much todo in Linux.

Can we now remove both workarounds `use_dark_theme` and `use_light_icon`, @Hofer-Julian ? One was needed for Gnome IIRC.

Also made a [screengrab](https://files.qmax.us/vorta-2020-05-984afb6096363eaec0dcd0a91725580aa7673f88.mp4
) 🤓

**Comment by Hofer-Julian** [comment]
The light tray icon should be the one to use for GNOME since the bar is black, yes.

Not sure about other DEs. Maybe we make it light per default and make it smarter as soon as people complain?

**Comment by m3nu** [comment]
Then I just add an override for Gnome and disable the option for now.

So the icon color will be based on the color palette in use. Except for Gnome, where it will be always white.

**Comment by Hofer-Julian** [comment]
Oh just saw your nice screencast.

KDE changes tray color with theme change.

Then probably the best option is to leave light icon hardcoded for GNOME, and see if icon change with theme change works for other DEs as well

**Comment by sten0** [comment]
Manu <notifications@github.com> writes:

> Thanks for the VM instructions. I was afraid to ask 😄. Will try that this morning.
>

You're welcome :-)  No problem!

> The other items are probably taken care of already. Detecting dark mode based on palette colors works perfectly and the style issue in `QToolbox` was due to a bold style I applied. After I removed it, it worked flawlessly. This may fix the titles as well, since the style was applied there too. (just to make it bold, but seems to confuse Qt)
>
> I'll check back after testing on the VM.

Manu <notifications@github.com> writes:

> Looks like all the issues are already solved. Not much todo in Linux.
>
> Can we now remove both workarounds `use_dark_theme` and `use_light_icon`, @Hofer-Julian ? One was needed for Gnome IIRC.
>
> Also made a [screengrab](https://files.qmax.us/vorta-2020-05-984afb6096363eaec0dcd0a91725580aa7673f88.mp4
> ) 🤓
>

Nice, I'm happy to see those instructions were sufficiently complete :-)
If you think they would be useful to other people, please let me know
and I can write up a HOWTO/Tutorial.

Cheers,
Nicholas
